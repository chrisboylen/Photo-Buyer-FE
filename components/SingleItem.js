import React, { Component } from "react";
import gql from "graphql-tag";
import { Query } from "react-apollo";
import Error from "./ErrorMessage";
import styled from "styled-components";
import Head from "next/head";
import formatMoney from "../lib/formatMoney";

const SingleItemStyles = styled.div`
  border-radius: 5px;
  max-width: 1200px;
  margin: 1rem auto;
  box-shadow: ${props => props.theme.bs};
  display: grid;
  grid-auto-columns: 1fr;
  grid-auto-flow: column;
  min-height: 800px;
  h2 {
    background: ${props => props.theme.blue};
    border-radius: 5px 5px 0 0;
    color: ${props => props.theme.white};
    text-align: center;
    margin: 0;
    span {
      color: #c7e4ff;
    }
  }
  img {
    width: 100%;
    height: 100%;
    object-fit: contain;
    border-radius: 0 0 5px 5px;
  }
  .details {
    margin: 3rem;
    font-size: 2rem;
  }
`;

const SINGLE_ITEM_QUERY = gql`
  query SINGLE_ITEM_QUERY($id: ID!) {
    item(where: { id: $id }) {
      id
      title
      description
      largeImage
    }
  }
`;
class SingleItem extends Component {
  render() {
    return (
      <Query
        query={SINGLE_ITEM_QUERY}
        variables={{
          id: this.props.id
        }}
      >
        {({ error, loading, data }) => {
          if (error) return <Error error={error} />;
          if (loading) return <p>Loading...</p>;
          if (!data.item) return <p>No Item Found for {this.props.id}</p>;
          const item = data.item;
          return (
            <SingleItemStyles>
              <Head>
                <title>Photo Buyer | {item.title}</title>
              </Head>
              <div className="details">
                <h2>
                  <span>Viewing: </span> {item.title}
                </h2>
                <img src={item.largeImage} alt={item.title} />
                <p>{item.description}</p>
              </div>
            </SingleItemStyles>
          );
        }}
      </Query>
    );
  }
}

export default SingleItem;
export { SINGLE_ITEM_QUERY };
