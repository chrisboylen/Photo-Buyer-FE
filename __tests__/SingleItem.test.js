import SingleItem, { SINGLE_ITEM_QUERY } from "../components/SingleItem";
import { shallow, mount } from "enzyme";
import toJSON from "enzyme-to-json";
import wait from "waait";
import { MockedProvider } from "react-apollo/test-utils";
import { fakeItem } from "../lib/testUtils";

describe("SINGLEITEM", () => {
  it("renders with proper data", async () => {
    const mocks = [
      {
        //when someone makes a req with this query and var combo
        request: { query: SINGLE_ITEM_QUERY, variables: { id: "123" } }
        //return this fake data (mocked data)
      }
    ];
    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <SingleItem id="123" />
      </MockedProvider>
    );

    expect(wrapper.text()).toContain("Loading...");
    wrapper.update();
    await wait();
    expect(toJSON(wrapper.find("h2"))).toMatchSnapshot();
  });

  it("errors with a not found item", async () => {
    const mocks = [
      {
        request: { query: SINGLE_ITEM_QUERY, variables: { id: "123" } },
        result: { errors: [{ message: "Items Not Found!" }] }
      }
    ];
    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <SingleItem id="123" />
      </MockedProvider>
    );
    await wait();
    wrapper.update();
    const item = wrapper.find('[data-test="graphql-error"]');
    expect(item.text()).toContain("Items Not Found!");
    expect(toJSON(item)).toMatchSnapshot();
  });
});
