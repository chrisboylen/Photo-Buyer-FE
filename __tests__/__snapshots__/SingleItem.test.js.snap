// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SINGLEITEM errors with a not found item 1`] = `
<p
  data-test="graphql-error"
>
  <strong>
    Shoot!
  </strong>
  Items Not Found!
</p>
`;

exports[`SINGLEITEM renders with proper data 1`] = `null`;
